# Читання даних з веб-сторінок

---

Для початку варто інсталювати пакет `rvest`.

In [102]:
# install.packages("rvest")

Завантажуємо пакет у память, щоб можна було використовувати функції.

In [103]:
library(rvest)

Вказуємо адресу ресурсу для читання даних:

In [104]:
url <- "https://abit-poisk.org.ua/rate2022/direction/968614"
url

[1] "https://abit-poisk.org.ua/rate2022/direction/968614"

Читаємо сторінку у змінну у форматі `html` як набір тексту:

In [105]:
page <- read_html(url)
page

{html_document}
<html lang="uk">
[1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8 ...
[2] <body>\n<div id="js-ba-options" data-options='{"banners":[["\\/assets\\/i ...

Витягуємо з інформації про сторінку усі потрібні нам теги, у даному випадку це таблиці, тег `table`.

In [106]:
tables <- html_nodes(page, "table")
tables

{xml_nodeset (2)}
[1] <table class="table table-bordered table-hover">\n<thead><tr>\n<th class= ...
[2] <table class="table table-bordered table-hover" style="padding-right: 0;p ...

Конвертуємо отриману таблицю з формату html у `data.frame` та переглядаємо перших кілька рядків таблиці, щоб переконатися, що усе `ОК`.

In [112]:
data <- as.data.frame(html_table(tables[1]), fill = T)
data |> head()

X.
1 1 
2 2 
3 3 
4 4 
5 5 
6 6 
  ПІБ                                                                                                                                                                                                                    
1 1\n                                                        \n                                Харечко Ю. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1  
2 2\n                                                        \n                                Мосійчук Д. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
3 3\n                                                        \n                                Буднік Н. Р.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
4 4\n                                                        \n                                Требухова В. І.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 5
5 5\n                                                        \n                                Ясковець М. А.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
6 6\n                                                        \n                                Лепська Л. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2  
  П Заг.бал Статус       
1 1 189.45  До наказу (б)
2 1 186.00  До наказу (б)
3 1 184.70  До наказу (к)
4 5 178.90  До наказу (к)
5 1 178.80  До наказу (к)
6 2 171.45  До наказу (к)
  Складові.заг..балу                                                                                                                                                                                                                                                                                                                                          
1 187  185  200                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
2 200  180  176                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 180\n\n                                                                                    Історія України: 176
3 194  192  160                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 160
4 194  180  156                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 180\n\n                                                                                    Історія України: 156
5 184  176  176                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 176
6 184  172  153                                \n                                \n                                  

Підключимо бібліотеку dplyr для виконання вибірок та маніпулювання даними:

In [114]:
library(dplyr)

Видалимо 3-й та 7-й стовпець з інформацією про квоту

In [116]:
data <- data |> select(-c(3,7))
data |> head()

X.
1 1 
2 2 
3 3 
4 4 
5 5 
6 6 
  ПІБ                                                                                                                                                                                                                    
1 1\n                                                        \n                                Харечко Ю. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1  
2 2\n                                                        \n                                Мосійчук Д. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
3 3\n                                                        \n                                Буднік Н. Р.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
4 4\n                                                        \n                                Требухова В. І.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 5
5 5\n                                                        \n                                Ясковець М. А.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
6 6\n                                                        \n                                Лепська Л. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2  
  Заг.бал Статус       
1 189.45  До наказу (б)
2 186.00  До наказу (б)
3 184.70  До наказу (к)
4 178.90  До наказу (к)
5 178.80  До наказу (к)
6 171.45  До наказу (к)
  Складові.заг..балу                                                                                                                                                                                                                                                                                                                                          
1 187  185  200                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
2 200  180  176                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 180\n\n                                                                                    Історія України: 176
3 194  192  160                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 160
4 194  180  156                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 180\n\n                                                                                    Історія України: 156
5 184  176  176                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 176
6 184  172  153                                \n                                \n                                    Українська м

Замінимо назви стовпців дата-фрейму для зручності використання під час роботи з даними

In [117]:
colnames(data) <- c("Rank", "Name", "TotalGrade", "StudyType", "Grades", "IsStudent")
data |> head()

Rank
1 1   
2 2   
3 3   
4 4   
5 5   
6 6   
  Name                                                                                                                                                                                                                   
1 1\n                                                        \n                                Харечко Ю. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1  
2 2\n                                                        \n                                Мосійчук Д. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
3 3\n                                                        \n                                Буднік Н. Р.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
4 4\n                                                        \n                                Требухова В. І.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 5
5 5\n                                                        \n                                Ясковець М. А.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
6 6\n                                                        \n                                Лепська Л. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2  
  TotalGrade StudyType    
1 189.45     До наказу (б)
2 186.00     До наказу (б)
3 184.70     До наказу (к)
4 178.90     До наказу (к)
5 178.80     До наказу (к)
6 171.45     До наказу (к)
  Grades                                                                                                                                                                                                                                                                                                                                                      
1 187  185  200                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
2 200  180  176                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 180\n\n                                                                                    Історія України: 176
3 194  192  160                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 160
4 194  180  156                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 180\n\n                                                                                    Історія України: 156
5 184  176  176                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 176
6 184  172  153                                \n                                \n             

Підготуємо наш набір даних до вигляду, коли у нас будуть стовпці у вигляді

|Rank|Name|StudyType|UkrGrade|MathGrade|HistoryGrade|TotalGrade|IsStudent|
|---|---|---|---|---|---|---|---|
|1|Прізвище І.І.|Бюджет|158|198|184|172.4|TRUE|

`Rank` лишаємо без змін.

Очистимо стовпець `Name`. Для початку переглянемо, що насправді може бути у цьому стовпці:

In [119]:
data$Name[1]

[1] "1\n                                                        \n                                Харечко Ю. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1"

З чого він складається? Цифра на початку, `\n` та `\t`, багато зайвих пробілів та інформація про пріоритет вкінці  `Пр.`. 

Для початку позбудемося `\n` та `\t`.

In [121]:
library(stringr)

In [124]:
name <- data$Name[1]
name <- str_replace_all(name, "\n", "")
name

[1] "1                                                                                        Харечко Ю. В.                                                        \t\t\t\t\t\t                                  Пр. 1"

In [126]:
name <- str_replace_all(name, "\t", "")
name

[1] "1                                                                                        Харечко Ю. В.                                                                                          Пр. 1"

Тепер позбудемося пробілів, наприклад, замінимо 2 пробіли на 1.

In [128]:
str_replace_all(name, "  ", " ")

[1] "1                                            Харечко Ю. В.                                             Пр. 1"

Стало менше проте ми їх не позбулися (((

In [130]:
name <- str_replace_all(name, "  ", " ")
name <- str_replace_all(name, "  ", " ")
name

[1] "1      Харечко Ю. В.      Пр. 1"

Майже, але ще не все

In [132]:
name <- str_replace_all(name, "  ", " ")
name <- str_replace_all(name, "  ", " ")
name

[1] "1 Харечко Ю. В. Пр. 1"

Вийшло, проте ми не можемо гарантувати, що усі дані мають таку ж кількість пробілів. Варто написати алгоритм, що буде видаляти усі пробіли за 1 раз.

In [136]:
text_length = 0
text = name

# доки довжина тексту змінюється
while(text_length != nchar(text)) {
    # рахуємо скільки символів у тексті
    text_length <- nchar(text)    
    # замінюємо 2 пробіли на 1
    text <- str_replace_all(text, "  ", " ")   
}
text

[1] "1 Харечко Ю. В. Пр. 1"

Далі загорнемо це все у функцію:

In [140]:
text_cleaner <- function(text) {
    
    text <- str_replace_all(text, "\t", "")
    text <- str_replace_all(text, "\n", "")
    
    text_length = 0

    while(text_length != nchar(text)) {
        text_length <- nchar(text)    
        text <- str_replace_all(text, "  ", " ")   
    }
    text
}

Перевіримо чи функція працює:

In [142]:
text_cleaner(data$Name[3])

[1] "3 Буднік Н. Р. Пр. 1"

In [168]:
std_names <- sapply(data$Name, text_cleaner)
names(std_names) <- NULL
std_names

[1] "1 Харечко Ю. В. Пр. 1"       "2 Мосійчук Д. С. Пр. 1"     
 [3] "3 Буднік Н. Р. Пр. 1"        "4 Требухова В. І. Пр. 5"    
 [5] "5 Ясковець М. А. Пр. 1"      "6 Лепська Л. О. Пр. 2"      
 [7] "7 Гнатенко А. С. Пр. 1"      "8 Бажанова Б. О. Пр. 2"     
 [9] "9 Рибак К. О. Пр. 1"         "10 Сукач А. М. К"           
[11] "11 Фалько М. Д. К"           "12 Нечипорук В. М. Пр. 1"   
[13] "13 Кухарчук Д. В. Пр. 1"     "14 120-7996285 Б. М. Пр. 3" 
[15] "15 Куждіна У. О. Пр. 3"      "16 Синєгубова О. Є. Пр. 5"  
[17] "17 Серветник Є. М. Пр. 4"    "18 Зінгер А. В. Пр. 2"      
[19] "19 Кляпчук В. Ю. Пр. 2"      "20 Єрмакович О. Ю. Пр. 2"   
[21] "21 Ступницька С. П. Пр. 4"   "22 Шмигельська А. І. Пр. 1" 
[23] "23 Трофімчук А. І. Пр. 1"    "24 Корень В. А. Пр. 4"      
[25] "25 Кушнір І. М. Пр. 2"       "26 Марків Л. А. Пр. 4"      
[27] "27 Бабак У. О. К"            "28 Вакулка Н. А. Пр. 3"     
[29] "29 Літвінчук В. І. Пр. 2"    "30 Яроменко М. А. Пр. 5"    
[31] "31 Андросова Г. В. Пр. 1"    "32 Махлай А. М. Пр. 4"      
[33] "33 Мулярчук М. Д. Пр. 2"     "34 Лисковець М. М. Пр. 3"   
[35] "35 Негода Т. Ю. Пр. 3"       "36 Пясецька Н. О. Пр. 4"    
[37] "37 Березніцький Є. О. Пр. 3" "38 Жученя Т. В. Пр. 2"      
[39] "39 Луцкова В. В. К"          "40 Андрощук М. О. К"        
[41] "41 Захарченко Н. К. Пр. 5"   "42 Тимошицька В. В. К"      
[43] "43 Черняк К. В. Пр. 2"       "44 Бобер І. С. Пр. 5"       
[45] "45 Ліщук А. І. Пр. 5"        "46 Ремезюк В. О. Пр. 2"     
[47] "47 Пінчаківська В. А. Пр. 4" "48 Кондратюк Б. С. Пр. 2"   
[49] "49 Ошейко В. О. Пр. 1"       "50 Коцюк О. Ю. Пр. 2"       
[51] "51 Круш І. В. Пр. 2"         "52 Шапор Ю. О. Пр. 1"       
[53] "53 Дімітрова А. О. Пр. 1"    "54 Скоцький Т. А. Пр. 1"    
[55] "55 Вознюк А. В. Пр. 5"       "56 Квецко В. А. Пр. 4"      
[57] "57 Горлачук В. А. Пр. 2"     "58 Шуль М. В. Пр. 3"        
[59] "59 Калітко А. В. Пр. 2"

In [9]:
table1 <- table1 |>
    mutate(IsStudentNow = ifelse(IsStudentNow == "+", T, F))

In [10]:
as.vector(table1$Status)

[1] "До наказу (б)"             "До наказу (б)"            
 [3] "До наказу (к)"             "До наказу (к)"            
 [5] "До наказу (к)"             "До наказу (к)"            
 [7] "До наказу (к)"             "До наказу (к)"            
 [9] "До наказу (к)"             "До наказу (к)"            
[11] "До наказу (к)"             "До наказу (к)"            
[13] "До наказу (к)"             "Допущено"                 
[15] "Допущено"                  "Допущено"                 
[17] "Допущено"                  "Допущено"                 
[19] "Допущено"                  "Допущено"                 
[21] "Допущено"                  "Допущено"                 
[23] "Допущено"                  "Допущено"                 
[25] "Допущено"                  "Допущено"                 
[27] "Допущено"                  "Допущено"                 
[29] "Допущено"                  "Допущено"                 
[31] "Допущено"                  "Допущено"                 
[33] "Допущено"                  "Допущено"                 
[35] "Допущено"                  "Допущено"                 
[37] "Допущено"                  "Допущено"                 
[39] "Допущено"                  "Допущено"                 
[41] "Допущено"                  "Допущено"                 
[43] "Допущено"                  ""                         
[45] ""                          ""                         
[47] ""                          ""                         
[49] ""                          ""                         
[51] ""                          ""                         
[53] ""                          ""                         
[55] ""                          ""                         
[57] ""                          "Скасовано (втрата пріор.)"
[59] ""

In [11]:
table1 <- table1 |>
    mutate(Status = ifelse(Status == "До наказу (б)", "Бюджет",
                           ifelse(Status == "До наказу (к)", "Контракт",
                                  "Не навчається"))       
)

In [12]:
library(stringr)

In [13]:
ukr_grades <- c()
math_grades <- c()
history_grades <- c()

for(i in 1:nrow(table1)) {
    # 2 пробіли у тексті
    grades <- as.numeric(unlist(strsplit(str_replace_all(table1[i, ]$Grades, "  ", " "), " "))[1:3])
    ukr_grades <- c(ukr_grades, grades[1])
    math_grades <- c(math_grades, grades[2])
    history_grades <- c(history_grades, grades[3])
    }
table1 <- table1 |>
    mutate(UkrGrade = ukr_grades,
          MathGrade = math_grades,
          HistoryGrade = history_grades) |>
    select(-Grades)

In [151]:
cleaner <- function(text) {
    
    text <- str_replace_all(text, "\n", "")
    text <- str_replace_all(text, "\t", "")
    
    txt_length = 0;
    while(txt_length != nchar(text)) {        
        txt_length = nchar(text)
        text <- str_replace_all(text, "  ", " ")
    }
    
    return (text)
}

[1] "1 Харечко Ю. В. Пр. 1"       "2 Мосійчук Д. С. Пр. 1"     
 [3] "3 Буднік Н. Р. Пр. 1"        "4 Требухова В. І. Пр. 5"    
 [5] "5 Ясковець М. А. Пр. 1"      "6 Лепська Л. О. Пр. 2"      
 [7] "7 Гнатенко А. С. Пр. 1"      "8 Бажанова Б. О. Пр. 2"     
 [9] "9 Рибак К. О. Пр. 1"         "10 Сукач А. М. К"           
[11] "11 Фалько М. Д. К"           "12 Нечипорук В. М. Пр. 1"   
[13] "13 Кухарчук Д. В. Пр. 1"     "14 120-7996285 Б. М. Пр. 3" 
[15] "15 Куждіна У. О. Пр. 3"      "16 Синєгубова О. Є. Пр. 5"  
[17] "17 Серветник Є. М. Пр. 4"    "18 Зінгер А. В. Пр. 2"      
[19] "19 Кляпчук В. Ю. Пр. 2"      "20 Єрмакович О. Ю. Пр. 2"   
[21] "21 Ступницька С. П. Пр. 4"   "22 Шмигельська А. І. Пр. 1" 
[23] "23 Трофімчук А. І. Пр. 1"    "24 Корень В. А. Пр. 4"      
[25] "25 Кушнір І. М. Пр. 2"       "26 Марків Л. А. Пр. 4"      
[27] "27 Бабак У. О. К"            "28 Вакулка Н. А. Пр. 3"     
[29] "29 Літвінчук В. І. Пр. 2"    "30 Яроменко М. А. Пр. 5"    
[31] "31 Андросова Г. В. Пр. 1"    "32 Махлай А. М. Пр. 4"      
[33] "33 Мулярчук М. Д. Пр. 2"     "34 Лисковець М. М. Пр. 3"   
[35] "35 Негода Т. Ю. Пр. 3"       "36 Пясецька Н. О. Пр. 4"    
[37] "37 Березніцький Є. О. Пр. 3" "38 Жученя Т. В. Пр. 2"      
[39] "39 Луцкова В. В. К"          "40 Андрощук М. О. К"        
[41] "41 Захарченко Н. К. Пр. 5"   "42 Тимошицька В. В. К"      
[43] "43 Черняк К. В. Пр. 2"       "44 Бобер І. С. Пр. 5"       
[45] "45 Ліщук А. І. Пр. 5"        "46 Ремезюк В. О. Пр. 2"     
[47] "47 Пінчаківська В. А. Пр. 4" "48 Кондратюк Б. С. Пр. 2"   
[49] "49 Ошейко В. О. Пр. 1"       "50 Коцюк О. Ю. Пр. 2"       
[51] "51 Круш І. В. Пр. 2"         "52 Шапор Ю. О. Пр. 1"       
[53] "53 Дімітрова А. О. Пр. 1"    "54 Скоцький Т. А. Пр. 1"    
[55] "55 Вознюк А. В. Пр. 5"       "56 Квецко В. А. Пр. 4"      
[57] "57 Горлачук В. А. Пр. 2"     "58 Шуль М. В. Пр. 3"        
[59] "59 Калітко А. В. Пр. 2"

In [149]:
nchar(data$Name)

[1] 208 209 207 210 209 208 209 209 206 203 204 211 210 214 209 212 211 208 209
[20] 211 212 213 211 208 208 208 203 209 211 210 211 208 210 211 208 210 214 208
[39] 205 206 212 208 208 207 207 209 214 211 208 207 206 207 211 210 208 208 210
[58] 206 209

In [32]:
head(table1)

Rank Name                    Priority TotalGrade Status   IsStudentNow
1 1    1 Харечко Ю. В. Пр. 1   1        189.45     Бюджет   TRUE        
2 2    2 Мосійчук Д. С. Пр. 1  1        186.00     Бюджет   TRUE        
3 3    3 Буднік Н. Р. Пр. 1    1        184.70     Контракт TRUE        
4 4    4 Требухова В. І. Пр. 5 5        178.90     Контракт TRUE        
5 5    5 Ясковець М. А. Пр. 1  1        178.80     Контракт TRUE        
6 6    6 Лепська Л. О. Пр. 2   2        171.45     Контракт TRUE        
  UkrGrade MathGrade HistoryGrade
1 187      185       200         
2 200      180       176         
3 194      192       160         
4 194      180       156         
5 184      176       176         
6 184      172       153

In [39]:
indexes <- as.data.frame(str_locate(table1$Name, "Пр."))
head(indexes, 10)

start end
1  17    19 
2  18    20 
3  16    18 
4  19    21 
5  18    20 
6  17    19 
7  18    20 
8  18    20 
9  15    17 
10 NA    NA

Додати пояснення + поетапно

In [42]:
names <- ifelse(is.na(str_sub(table1$Name, start = 1L, end = indexes$start-2)), table1$Name, str_sub(table1$Name, start = 1L, end = indexes$start-2))

In [43]:
names

[1] "1 Харечко Ю. В."       "2 Мосійчук Д. С."      "3 Буднік Н. Р."       
 [4] "4 Требухова В. І."     "5 Ясковець М. А."      "6 Лепська Л. О."      
 [7] "7 Гнатенко А. С."      "8 Бажанова Б. О."      "9 Рибак К. О."        
[10] "10 Сукач А. М. К"      "11 Фалько М. Д. К"     "12 Нечипорук В. М."   
[13] "13 Кухарчук Д. В."     "14 120-7996285 Б. М."  "15 Куждіна У. О."     
[16] "16 Синєгубова О. Є."   "17 Серветник Є. М."    "18 Зінгер А. В."      
[19] "19 Кляпчук В. Ю."      "20 Єрмакович О. Ю."    "21 Ступницька С. П."  
[22] "22 Шмигельська А. І."  "23 Трофімчук А. І."    "24 Корень В. А."      
[25] "25 Кушнір І. М."       "26 Марків Л. А."       "27 Бабак У. О. К"     
[28] "28 Вакулка Н. А."      "29 Літвінчук В. І."    "30 Яроменко М. А."    
[31] "31 Андросова Г. В."    "32 Махлай А. М."       "33 Мулярчук М. Д."    
[34] "34 Лисковець М. М."    "35 Негода Т. Ю."       "36 Пясецька Н. О."    
[37] "37 Березніцький Є. О." "38 Жученя Т. В."       "39 Луцкова В. В. К"   
[40] "40 Андрощук М. О. К"   "41 Захарченко Н. К."   "42 Тимошицька В. В. К"
[43] "43 Черняк К. В."       "44 Бобер І. С."        "45 Ліщук А. І."       
[46] "46 Ремезюк В. О."      "47 Пінчаківська В. А." "48 Кондратюк Б. С."   
[49] "49 Ошейко В. О."       "50 Коцюк О. Ю."        "51 Круш І. В."        
[52] "52 Шапор Ю. О."        "53 Дімітрова А. О."    "54 Скоцький Т. А."    
[55] "55 Вознюк А. В."       "56 Квецко В. А."       "57 Горлачук В. А."    
[58] "58 Шуль М. В."         "59 Калітко А. В."

In [81]:
cleaner <- function(text) {
    text <- str_replace_all(text, "  ", " ")
    text <- str_replace_all(text, "\n", "")
    text <- str_replace_all(text, "\t", "")
    return (text)
    }

In [71]:
x <- unlist(str_split(names[1], ' '))

In [72]:
class(x)

[1] "character"

In [73]:
x[2:length(x)]

[1] "Харечко" "Ю."      "В."

In [81]:
paste(x[2:length(x)], collapse = " ")

[1] "Харечко Ю. В."

In [82]:
cleaner(table1$Name)

Warning message in while (txt_length != nchar(text)) {:
"the condition has length > 1 and only the first element will be used"
Warning message in while (txt_length != nchar(text)) {:
"the condition has length > 1 and only the first element will be used"


[1] "1 Харечко Ю. В. Пр. 1"       "2 Мосійчук Д. С. Пр. 1"     
 [3] "3 Буднік Н. Р. Пр. 1"        "4 Требухова В. І. Пр. 5"    
 [5] "5 Ясковець М. А. Пр. 1"      "6 Лепська Л. О. Пр. 2"      
 [7] "7 Гнатенко А. С. Пр. 1"      "8 Бажанова Б. О. Пр. 2"     
 [9] "9 Рибак К. О. Пр. 1"         "10 Сукач А. М. К"           
[11] "11 Фалько М. Д. К"           "12 Нечипорук В. М. Пр. 1"   
[13] "13 Кухарчук Д. В. Пр. 1"     "14 120-7996285 Б. М. Пр. 3" 
[15] "15 Куждіна У. О. Пр. 3"      "16 Синєгубова О. Є. Пр. 5"  
[17] "17 Серветник Є. М. Пр. 4"    "18 Зінгер А. В. Пр. 2"      
[19] "19 Кляпчук В. Ю. Пр. 2"      "20 Єрмакович О. Ю. Пр. 2"   
[21] "21 Ступницька С. П. Пр. 4"   "22 Шмигельська А. І. Пр. 1" 
[23] "23 Трофімчук А. І. Пр. 1"    "24 Корень В. А. Пр. 4"      
[25] "25 Кушнір І. М. Пр. 2"       "26 Марків Л. А. Пр. 4"      
[27] "27 Бабак У. О. К"            "28 Вакулка Н. А. Пр. 3"     
[29] "29 Літвінчук В. І. Пр. 2"    "30 Яроменко М. А. Пр. 5"    
[31] "31 Андросова Г. В. Пр. 1"    "32 Махлай А. М. Пр. 4"      
[33] "33 Мулярчук М. Д. Пр. 2"     "34 Лисковець М. М. Пр. 3"   
[35] "35 Негода Т. Ю. Пр. 3"       "36 Пясецька Н. О. Пр. 4"    
[37] "37 Березніцький Є. О. Пр. 3" "38 Жученя Т. В. Пр. 2"      
[39] "39 Луцкова В. В. К"          "40 Андрощук М. О. К"        
[41] "41 Захарченко Н. К. Пр. 5"   "42 Тимошицька В. В. К"      
[43] "43 Черняк К. В. Пр. 2"       "44 Бобер І. С. Пр. 5"       
[45] "45 Ліщук А. І. Пр. 5"        "46 Ремезюк В. О. Пр. 2"     
[47] "47 Пінчаківська В. А. Пр. 4" "48 Кондратюк Б. С. Пр. 2"   
[49] "49 Ошейко В. О. Пр. 1"       "50 Коцюк О. Ю. Пр. 2"       
[51] "51 Круш І. В. Пр. 2"         "52 Шапор Ю. О. Пр. 1"       
[53] "53 Дімітрова А. О. Пр. 1"    "54 Скоцький Т. А. Пр. 1"    
[55] "55 Вознюк А. В. Пр. 5"       "56 Квецко В. А. Пр. 4"      
[57] "57 Горлачук В. А. Пр. 2"     "58 Шуль М. В. Пр. 3"        
[59] "59 Калітко А. В. Пр. 2"

In [83]:
str_replace_all(str_replace_all(table1$Name, "  ", " "), "\n", "")

[1] "1 Харечко Ю. В. Пр. 1"       "2 Мосійчук Д. С. Пр. 1"     
 [3] "3 Буднік Н. Р. Пр. 1"        "4 Требухова В. І. Пр. 5"    
 [5] "5 Ясковець М. А. Пр. 1"      "6 Лепська Л. О. Пр. 2"      
 [7] "7 Гнатенко А. С. Пр. 1"      "8 Бажанова Б. О. Пр. 2"     
 [9] "9 Рибак К. О. Пр. 1"         "10 Сукач А. М. К"           
[11] "11 Фалько М. Д. К"           "12 Нечипорук В. М. Пр. 1"   
[13] "13 Кухарчук Д. В. Пр. 1"     "14 120-7996285 Б. М. Пр. 3" 
[15] "15 Куждіна У. О. Пр. 3"      "16 Синєгубова О. Є. Пр. 5"  
[17] "17 Серветник Є. М. Пр. 4"    "18 Зінгер А. В. Пр. 2"      
[19] "19 Кляпчук В. Ю. Пр. 2"      "20 Єрмакович О. Ю. Пр. 2"   
[21] "21 Ступницька С. П. Пр. 4"   "22 Шмигельська А. І. Пр. 1" 
[23] "23 Трофімчук А. І. Пр. 1"    "24 Корень В. А. Пр. 4"      
[25] "25 Кушнір І. М. Пр. 2"       "26 Марків Л. А. Пр. 4"      
[27] "27 Бабак У. О. К"            "28 Вакулка Н. А. Пр. 3"     
[29] "29 Літвінчук В. І. Пр. 2"    "30 Яроменко М. А. Пр. 5"    
[31] "31 Андросова Г. В. Пр. 1"    "32 Махлай А. М. Пр. 4"      
[33] "33 Мулярчук М. Д. Пр. 2"     "34 Лисковець М. М. Пр. 3"   
[35] "35 Негода Т. Ю. Пр. 3"       "36 Пясецька Н. О. Пр. 4"    
[37] "37 Березніцький Є. О. Пр. 3" "38 Жученя Т. В. Пр. 2"      
[39] "39 Луцкова В. В. К"          "40 Андрощук М. О. К"        
[41] "41 Захарченко Н. К. Пр. 5"   "42 Тимошицька В. В. К"      
[43] "43 Черняк К. В. Пр. 2"       "44 Бобер І. С. Пр. 5"       
[45] "45 Ліщук А. І. Пр. 5"        "46 Ремезюк В. О. Пр. 2"     
[47] "47 Пінчаківська В. А. Пр. 4" "48 Кондратюк Б. С. Пр. 2"   
[49] "49 Ошейко В. О. Пр. 1"       "50 Коцюк О. Ю. Пр. 2"       
[51] "51 Круш І. В. Пр. 2"         "52 Шапор Ю. О. Пр. 1"       
[53] "53 Дімітрова А. О. Пр. 1"    "54 Скоцький Т. А. Пр. 1"    
[55] "55 Вознюк А. В. Пр. 5"       "56 Квецко В. А. Пр. 4"      
[57] "57 Горлачук В. А. Пр. 2"     "58 Шуль М. В. Пр. 3"        
[59] "59 Калітко А. В. Пр. 2"

In [64]:
table1 |> head(15)

Rank
1   1  
2   2  
3   3  
4   4  
5   5  
6   6  
7   7  
8   8  
9   9  
10 10  
11 11  
12 12  
13 13  
14 14  
15 15  
   Name                                                                                                                                                                                                                       
1  1\n                                                        \n                                Харечко Ю. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1      
2  2\n                                                        \n                                Мосійчук Д. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
3  3\n                                                        \n                                Буднік Н. Р.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
4  4\n                                                        \n                                Требухова В. І.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 5    
5  5\n                                                        \n                                Ясковець М. А.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
6  6\n                                                        \n                                Лепська Л. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2      
7  7\n                                                        \n                                Гнатенко А. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
8  8\n                                                        \n                                Бажанова Б. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2     
9  9\n                                                        \n                                Рибак К. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1        
10 10\n                                                        \n                                Сукач А. М.                            \n\n                            \n\t\t\t\t\t\t                                  К           
11 11\n                                                        \n                                Фалько М. Д.                            \n\n                            \n\t\t\t\t\t\t                                  К          
12 12\n                                                        \n                                Нечипорук В. М.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
13 13\n                                                        \n                                Кухарчук Д. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
14 14\n                                                        \n                                120-7996285  Б. М.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3
15 15\n                                                        \n                                Куждіна У. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3     
   Priority TotalGrade Status        IsStudentNow UkrGrade MathGrade
1  1        189.45     Бюджет         TRUE        187      185      
2  1        186.00     Бюджет         TRUE        200      180      
3  1      

In [33]:
library(stringr)

In [36]:
table1$TotalAssets <- as.numeric(gsub(",","" , table1$TotalAssets))

In [37]:
mean(table1$TotalAssets)

[1] 1404.41

In [24]:
table2 <- as.data.frame(html_table(tables[2]), fill = T)
table2 |> head()

Rank Bank.name                                      
1 1    Industrial and Commercial Bank of China Limited
2 2    China Construction Bank                        
3 3    Agricultural Bank of China                     
4 4    Bank of China                                  
5 5    JPMorgan Chase                                 
6 6    Bank of America                                
  Total.assets.2022..US..billion.
1 5,742.86                       
2 5,016.81                       
3 4,919.03                       
4 4,192.12                       
5 3,736.62                       
6 3,100.02